<div style="border: solid blue 2px; padding: 20px; margin: 10px">

  <b>Overall Summary of the Project – Iteration 1</b><br><br>
  Hello Andy, Congratulations on submitting your project! 🎉<br><br>
  My name is <b>Victor Camargo</b> 
  (<a href="https://hub.tripleten.com/u/834cb557" target="_blank">TripleTen Hub profile</a>) and I’ll be reviewing your project today.<br><br>

  <b>Nice work on:</b><br>
  ✅ Proper data loading and inspection with clean structure and no missing values.<br>
  ✅ Clear and correct 60/20/20 split ensuring reproducibility.<br>
  ✅ Excellent comparison of models, well-reasoned interpretation, and detailed final evaluation with precision, recall, and accuracy metrics.<br><br>

  Your work is complete, consistent, and meets all project requirements. The explanations are clear, and you demonstrated a solid understanding of model selection, validation, and evaluation.<br><br>

  🟢 Project status: <b>Approved – great work!</b><br><br>

  <hr><b>Legend:</b><br>

  <div class="alert alert-success" style="border-left: 7px solid green; padding: 5px">
    <b>✅ Reviewer’s comment – Iteration 1:</b><br>
    Efficient solutions, strong ideas, or formatting done well. These can often be reused in future projects.
  </div>

  <div class="alert alert-warning" style="border-left: 7px solid gold; padding: 5px">
    <b>⚠️ Reviewer’s comment – Iteration 1:</b><br>
    Suggestions for optimisation. They are not mandatory but highly recommended to strengthen your work.
  </div>

  <div class="alert alert-danger" style="border-left: 7px solid red; padding: 5px">
    <b>⛔️ Reviewer’s comment – Iteration 1:</b><br>
    Indicates a problem that must be fixed. The project can’t be approved until this is resolved.
  </div>

  <div class="alert alert-info" style="border-left: 7px solid blue; padding: 5px">
    <b>🔵 Student’s Comment</b><br>
    To leave your own notes, questions, or explanations, create a <b>Markdown cell</b> in your notebook and copy-paste this code:<br>
    <code>&lt;div class="alert alert-info"; style="border-left: 7px solid blue"&gt;&lt;b&gt;Student’s Comment&lt;/b&gt;&lt;/div&gt;</code>
  </div>

  <hr>
  <b>Please ensure</b> that all cells run smoothly from top to bottom and display their outputs before submitting — this helps keep your analysis easy to follow.<br>
  <b>Kind reminder:</b> please do not remove or change reviewer comments, as they help track progress and make future reviews smoother.<br>
  <b>Feel free to reach out if you need help in the Questions channel.</b>
</div>


Intro to Machine Learning: Project
-

Project Description
--------------------
Mobile carrier Megaline has found out that many of their subscribers use legacy plans. They want to develop a model that would analyze subscribers' behavior and recommend one of Megaline's newer plans: Smart or Ultra. 

You have access to behavior data about subscribers who have already switched to the new plans (from the project for the Statistical Data Analysis course). For this classification task, you need to develop a model that will pick the right plan. Since you’ve already performed the data preprocessing step, you can move straight to creating the model.  

Develop a model with the highest possible accuracy. In this project, the threshold for accuracy is 0.75. Check the accuracy using the test dataset.  

Data Description
-----------------
Every observation in the dataset contains monthly behavior information about one user. The information given is as follows: 

сalls — number of calls

minutes — total call duration in minutes

messages — number of text messages

mb_used — Internet traffic used in MB

is_ultra — plan for the current month (Ultra - 1, Smart - 0)

In [2]:
#Loading packages and data
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score, classification_report

In [3]:
df = pd.read_csv('/datasets/users_behavior.csv')

In [4]:
#Checking dataframe and datasets
df.describe()
df.info()
df.isnull().sum()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

Looks like there are no missing data values in this dataset.

<div class="alert alert-success" style="border-left: 7px solid green; padding: 5px">
  <b>✅ Reviewer’s comment – Iteration 1:</b><br>
  Great start! All libraries were correctly imported, the dataset was loaded, and the initial inspection was done properly with no missing values detected.
</div>

<div class="alert alert-warning" style="border-left: 7px solid gold; padding: 5px">
  <b>⚠️ Reviewer’s comment – Iteration 1:</b><br>
  Consider adding a few simple charts, like histograms or a correlation heatmap, to visualize data distributions and patterns more clearly.
</div>


Data Spliting
--------------
We are looking to split ths source datga into a training set, a validation set, and a test set. Since we are spliting the data into 3 sets, the validation set and test set should be equal.  I will split 20% for each and 60% for training.

In [5]:
features = df.drop(['is_ultra'], axis=1)
target = df['is_ultra']
features_train, features_temp, target_train, target_temp = train_test_split(
    features, target, test_size=0.4, random_state=12345)
features_valid, features_test, target_valid, target_test = train_test_split(
    features_temp, target_temp, test_size=0.5, random_state=12345)

print(features_train.shape)
print(target_train.shape)
print(features_valid.shape)
print(target_valid.shape)
print(features_test.shape)
print(target_test.shape)

(1928, 4)
(1928,)
(643, 4)
(643,)
(643, 4)
(643,)


<div class="alert alert-success" style="border-left: 7px solid green; padding: 5px">
  <b>✅ Reviewer’s comment – Iteration 1:</b><br>
  Nice work splitting the dataset into training, validation, and test sets with a balanced 60/20/20 ratio. The use of <code>random_state</code> ensures reproducibility, and the code structure is clear and correct.
</div>

Model Studying and Hyperparameters Tuning
------------------------------------------
Now that we have split our data into 60/20/20.  I will begin to apply different models to determine which will be the best according to the findings.

Decision Tree Model
-

In [7]:
for depth in range (1, 6):
    model =  DecisionTreeClassifier(random_state = 12345, max_depth=depth)
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    print("max_depth =", depth)
    print(accuracy_score(target_valid, predictions_valid))

max_depth = 1
0.7542768273716952
max_depth = 2
0.7822706065318819
max_depth = 3
0.7853810264385692
max_depth = 4
0.7791601866251944
max_depth = 5
0.7791601866251944


We can see that at the max depth of 3, we have the highest accuracy. As we increase the depth, we can see the accuracy began to temporary drop and if we continue, we will see the result of overfitting.  

Random Forest Model
-

In [8]:
for est in range(1, 50):
    model = RandomForestClassifier(random_state=12345, n_estimators=est)
    model.fit(features_train, target_train)
    score = model.score(features_valid, target_valid)
    print("Accuracy of model on the validation set (n_estimators = {}): {}".format(est, score))

Accuracy of model on the validation set (n_estimators = 1): 0.7107309486780715
Accuracy of model on the validation set (n_estimators = 2): 0.7636080870917574
Accuracy of model on the validation set (n_estimators = 3): 0.7387247278382582
Accuracy of model on the validation set (n_estimators = 4): 0.7713841368584758
Accuracy of model on the validation set (n_estimators = 5): 0.749611197511664
Accuracy of model on the validation set (n_estimators = 6): 0.7807153965785381
Accuracy of model on the validation set (n_estimators = 7): 0.7682737169517885
Accuracy of model on the validation set (n_estimators = 8): 0.7822706065318819
Accuracy of model on the validation set (n_estimators = 9): 0.7729393468118196
Accuracy of model on the validation set (n_estimators = 10): 0.7853810264385692
Accuracy of model on the validation set (n_estimators = 11): 0.7838258164852255
Accuracy of model on the validation set (n_estimators = 12): 0.7869362363919129
Accuracy of model on the validation set (n_estimat

It takes about 10 estimators to get the a similar accuracy as our Decision Tree model.  The more estimators would mean more time needed for training.  As we continue to increase estimators, the accuracy does not continue to improve.

Logical Regression Model
-

In [9]:
model =  LogisticRegression(random_state=12345, solver='liblinear')
model.fit(features_train, target_train)
predictions_valid = model.predict(features_valid)
accuracy = model.score(features_valid, target_valid)
print("Accuracy of model =", accuracy)

Accuracy of model = 0.7091757387247278


Logical Regression model has the lowest accuracy between all 3 models and is < 75% which is under our threshold that we want for this project.

<div class="alert alert-success" style="border-left: 7px solid green; padding: 5px">
  <b>✅ Reviewer’s comment – Iteration 1:</b><br>
  Excellent comparison of models and hyperparameters. You correctly explored multiple depths for Decision Tree, tested a range of estimators for Random Forest, and included Logistic Regression for reference. Your explanations about overfitting and accuracy trends show good understanding of model behavior.
</div>

Model Quality Check with Test Set
-

In [10]:
best_model = RandomForestClassifier(random_state=12345, n_estimators=23)
best_model.fit(features_train, target_train)
predictions_test = best_model.predict(features_test)
final_accuracy = accuracy_score(target_test, predictions_test)
print("Final test accuracy:", final_accuracy)

print('Accuracy of the best model on the train set:', best_model.score(features_train, target_train))
print('Accuracy of the best model on the validation set:', best_model.score(features_valid, target_valid))
print('Accuracy of the best model on the test set:', best_model.score(features_test, target_test))

Final test accuracy: 0.7807153965785381
Accuracy of the best model on the train set: 0.9937759336099585
Accuracy of the best model on the validation set: 0.7947122861586314
Accuracy of the best model on the test set: 0.7807153965785381


In [11]:
precision = precision_score(best_model.predict(features_test), target_test)
print('Precision score of best model =', precision)
recall = recall_score(target_test, best_model.predict(features_test))
print('Recall score of best model =', recall)
print('\nDetailed classification report:')
print(classification_report(target_test, best_model.predict(features_test)))

Precision score of best model = 0.5812807881773399
Recall score of best model = 0.5812807881773399

Detailed classification report:
              precision    recall  f1-score   support

           0       0.82      0.87      0.84       440
           1       0.68      0.58      0.63       203

    accuracy                           0.78       643
   macro avg       0.75      0.73      0.74       643
weighted avg       0.77      0.78      0.78       643



Conclusions and Sanity Check
-
Between the 3 models, Logistic Regression have the lowest accuracy and was not able to meet our threshold of 75%.
Decision Tree model does not have an overfit problem at a max depth of 3.
Random Forest model is indeed overfitted but yields the highest accuracy.

Now just because we meet accuracy threshold for the project does not mean that the models are the best and completely applicable in real world predictions.  We noticed that the precision is quite low. With a precision of 58%, it means that 42% of customers that are recommended for Ultra either does not need it or is not utilizing it to its full capacity. This could lead to dissatisfication and frustration and could lose customers in the long term.  

From a model perspective, our model might be a bit too aggressive in suggesting the Ultra plan to customers.  THe current Random Forest model will most likely require much adjustment in which I did not do because it did not say it was required in the project. We could control depth to limit the growth of each tree as well as potentially adding max_features to add more randomness and reduce overfitting.  

If we are going with a more realistic approach, I would actually use Decision Tree model over Random Forest model because even though the accuracy is slightly lower with Decision Tree, it does not have an overfitting issue which would most likely lead to a higher precision score.

<div class="alert alert-success" style="border-left: 7px solid green; padding: 5px">
  <b>✅ Reviewer’s comment – Iteration 1:</b><br>
  Excellent evaluation of the final model. You clearly tested on the unseen data, reported all key metrics (accuracy, precision, recall, and classification report), and provided a thoughtful interpretation of the results. The discussion on trade-offs between accuracy and overfitting demonstrates solid analytical thinking.
</div>